### ISSUE TITLE
Condense iOS and macOS into a single Xcode project

### ISSUE THREAD

**Issue by leighajarett** [flutter/flutter#114179]
## Use case
It seems that when you deploy a Flutter app to iOS and macOS, you would need to take some repetitive steps. For example, opening the 'ios' folder in Xcode to change build settings, and then later open the 'macos' folder in Xcode to change the build settings again. 

Steps are detailed here: https://docs.flutter.dev/deployment/macos#review-xcode-project-settings

## Proposal
When you use the Swift multi-platform you have one project, with one target for the app itself. The project settings specify two deployment targets, and the app target allows you control the specifics for macOS and iOS as shown below. If Flutter generated just a single project, it would make it easier to complete some of the duplicate steps and also may make it easier to manage common assets for the app store. 

<img width="871" alt="Screenshot 2022-10-27 at 4 25 17 PM" src="https://user-images.githubusercontent.com/21065911/198391245-07f6cb84-feb0-47e0-af08-4aaf8408f440.png">

Full disclosure, I'm not even sure this is a good idea. If most people are just developing for iOS and Android, then this might give unnecessary overhead (deleting the mac target information). Another alternative is adding documentation that walks through adding macOS as a supported destination for the iOS runner project, but not sure if that would just work, or if will be confusing if the macOS project still exists.

**Comment by jmagman** [comment]
+1 it's really a bummer this isn't how it was designed from the beginning, and all the iOS and macOS tooling would be essentially identical, the biggest change just being what's targeted.  `Flutter.xcframework` could include all platforms including macOS.  Now that macOS is stable it would be really hard to merge the two together without breaking everything...

**Comment by stuartmorgan** [comment]
> it's really a bummer this isn't how it was designed from the beginning

It's not really clear to me how that would have worked in Flutter's model. Even ignoring the issue of desktop support being experimental and flag-controlled for quite a while (which would have been problematic if sharing a project), what would the path for "I have an existing Flutter app for mobile, and now I want to add macOS" be in that model? Maintain a full Xcode project parser so that we could inject new targets and settings into a project that we allow people to modify in near-arbitrary ways? Require people to throw away their iOS project and make a new project from the template? Describe how to update the project by hand? None of the options seem great.

> Now that macOS is stable it would be really hard to merge the two together without breaking everything...

If we have a reasonable design for a merged version we could potentially change the template, and have two codepaths. We're already sharing a bunch of code internally, and we may be able to share more even within the current structure, so it wouldn't be two full copies of all the code.

(We'd have to have done that to some extent regardless, unless we wanted to have macOS support be entirely managed from a folder called "ios" forever.)

**Comment by jmagman** [comment]
If they were both supported from the beginning, I would have designed the iOS and macOS projects to be the same (in a `darwin` folder? Wish there were a more colloquial word than `apple` for these platforms.).  You wouldn't be able to opt in to iOS without also getting macOS.  xcframework wasn't a thing back then, but if I designed it now there would be one `Flutter.xcframework` that supported macOS and iOS, it would be the same artifact.  There would be a single Xcode project with different macOS and iOS targets.  Plugins would by default support both, and it would be up to the plugin author to either `#if` the target conditionals, or set up their podspec to swap on the platform, if needed (linking AppKit vs UIKit, for example).

> Even ignoring the issue of desktop support being experimental and flag-controlled for quite a while (which would have been problematic if sharing a project)

That's the kicker, by "from the beginning" I meant if Flutter had supported macOS and iOS on day one.  Because macOS was added later this is theoretical.  I don't think there's a good path forward without requiring the user to recreate one or both Xcode projects, or the very tricky and probably error-prone Xcode project merge tool you describe.

**Comment by stuartmorgan** [comment]
If we were willing to maintain legacy `ios`, legacy `macos`, and combined `darwin`, it should be doable. We could just require that projects have either the legacy folder(s) or the combined folder, but not both. We'd need to do the xcframework, but updating existing macOS projects to the new name would be reasonable with the auto-migrator.

(Documenting it would be kind of a mess, but not impossible.)

I've considered something along these lines (a combined folder/protect option) for macOS/iOS plugins to eliminate the annoying symlinking approach.

**Comment by jmagman** [comment]
> We'd need to do the xcframework, but updating existing macOS projects to the new name would be reasonable with the auto-migrator.

Cross-referencing our conversation from https://github.com/flutter/flutter/issues/70413

**Comment by jmagman** [comment]
@a-wallen @cbracken

**Comment by a-wallen** [comment]
Neat idea. Before we get too concerned with the implementation details, do we know that people want this?

**Comment by a-wallen** [comment]
cc @melyndahoover, this could solve some pain points in the deployment of macOS apps.

**Comment by jmagman** [comment]
> do we know that people want this?

Even if there's minimal user impact, it would increase engineering velocity by minimizing the duplication between the iOS and macOS and Xcode logic in the tool.  Improvements like https://github.com/flutter/flutter/pull/113302 could be done once instead of either twice, or only on iOS thus further fragmenting the developer experience between the two platforms.  For example, you wouldn't be implementing flavors in macOS, you would just get all that logic for free 🙂

**Comment by a-wallen** [comment]
> For example, you wouldn't be implementing flavors in macOS, you would just get all that logic for free 🙂
SGTM 🤣

**Comment by a-wallen** [comment]
@jmagman do you think that there's any chance that having a multi-platform XCode project will be deprecated for some reason in the future?

**Comment by jmagman** [comment]
No it wouldn't get removed, you essentially have to do this if you want to say have a watch app and iOS app.  This Xcode multi-target pattern way predates the SwiftUI-friendly screenshot @leighajarett provided.

This can either be done with separate app targets:
<img width="154" alt="Screen Shot 2022-10-28 at 5 07 16 PM" src="https://user-images.githubusercontent.com/682784/198752480-bda455f0-a00a-4bae-a268-58d3dd1f7a82.png">
In which case you need to set up build settings twice in the same project, which @leighajarett is trying to avoid.

Or with one app targeting both iOS and macOS, but the code itself gets riddled with target checks:
```swift
import Cocoa

#if TARGET_OS_OSX
import FlutterMacOS

@NSApplicationMain
class AppDelegate: FlutterAppDelegate {
  override func applicationShouldTerminateAfterLastWindowClosed(_ sender: NSApplication) -> Bool {
    return true
  }
}

#else TARGET_OS_IOS

import Flutter

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}

#endif
```
The advantage of the latter is that there's only one app target to set up bundle identifiers, code signing, etc.  However that's also a disadvantage because often developers need different build settings between iOS and macOS, which is harder to do in the same target in Xcode's UI. 

Using SwiftUI in the template might help a bit with the target checking, but I suspect not since mostly the app is setting up `UIApplication` / `NSApplication` stuff.  And anyway SwiftUI only works on iOS 12 and higher, and our min is currently iOS 11.

I'm not sure either way would be much of a win from the user's perspective, and possibly it would confuse them more...

**Comment by stuartmorgan** [comment]
> it would increase engineering velocity by minimizing the duplication between the iOS and macOS and Xcode logic in the tool. Improvements like #113302 could be done once instead of either twice, or only on iOS thus further fragmenting the developer experience between the two platforms.

I do think we could get a lot of this benefit just by refactoring inside the tool, invisibly to users. I built the macOS tooling support as "add a new platform, sharing some code with iOS when it was obviously straightforward" rather than "aggressively generalize the existing iOS code to support working with either `ios/Runner*` or `macos/Runner*`, which in retrospect was a mistake. In my defense I was trying to make sure my macOS work didn't accidentally break iOS code, which I didn't have a deep end-to-end understanding of, but it means we have more divergence than we should even given a two-project structure.

If we invested some resources in going back and incrementally refactoring as much of the iOS and macOS tooling support as we can to share the same codepath, things like https://github.com/flutter/flutter/pull/113302 should become inherently shared even without fixing this issue.

**Comment by jmagman** [comment]
> I do think we could get a lot of this benefit just by refactoring inside the tool, invisibly to users.

That's true, but I was also thinking about the differences in the project structure, the extra `Flutter Assemble` aggregate target with the tripwire/xcfilelist inputs and outputs, the extra `AppInfo` and `Warning` xcconfigs.  It's subtly different and sometimes it's so different that I don't migrate macOS projects at the same time I'm migrating iOS because I need to get iOS out the door https://github.com/flutter/flutter/pull/51453 (I usually follow up though 🙂 https://github.com/flutter/flutter/pull/71764).  It's more work for sure.

**Comment by stuartmorgan** [comment]
Maybe we should file a new umbrella bug with a checklist of those differences? We could chip away at many of those too. E.g., the xcconfigs we could just decide which we like better and update the templates to match for new projects, since IIRC the tooling doesn't care about them. (IIRC, I made AppInfo.xcconfig because people used to have to regularly blow away and re-create the macOS template, which is no longer a factor, and because I thought it would be easier for people not familiar with Xcode to update core things, which is still relevant but would be as relevant on iOS and we should just pick one approach to use for both.)

**Comment by melyndahoover** [comment]
> Maybe we should file a new umbrella bug with a checklist of those differences? 
I agree. I am trying to understand how this will effect the user journey, but it is a bit difficult to keep track of the pros and cons of the different proposals in this thread.
